File: fabric/src/main/java/com/dreamdisplays/DreamDisplaysMod.java
package com.dreamdisplays;

import com.dreamdisplays.net.*;
import com.dreamdisplays.render.World;
import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.fabric.api.client.command.v2.ClientCommandRegistrationCallback;
import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientLifecycleEvents;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.client.rendering.v1.world.WorldRenderEvents;
import net.fabricmc.fabric.api.networking.v1.PayloadTypeRegistry;
import net.minecraft.client.Camera;
import net.minecraft.client.Minecraft;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import org.jspecify.annotations.NullMarked;

@NullMarked
public class DreamDisplaysMod implements ClientModInitializer, Mod {
    @Override
    public void onInitializeClient() {
        Initializer.onModInit(this);

        ClientCommandRegistrationCallback.EVENT.register((dispatcher, dedicated) -> dispatcher.register(
                LiteralArgumentBuilder.<FabricClientCommandSource>literal("displays")
                        .then(LiteralArgumentBuilder.<FabricClientCommandSource>literal("off")
                                .executes((context) -> {
                                    Initializer.displaysEnabled = false;
                                    return 1;
                                })
                        )
                        .then(LiteralArgumentBuilder.<FabricClientCommandSource>literal("on")
                                .executes((context) -> {
                                    Initializer.displaysEnabled = true;
                                    return 1;
                                })
                        )
        ));

        PayloadTypeRegistry.playS2C().register(Info.PACKET_ID, Info.PACKET_CODEC);

        PayloadTypeRegistry.playS2C().register(Sync.PACKET_ID, Sync.PACKET_CODEC);
        PayloadTypeRegistry.playC2S().register(Sync.PACKET_ID, Sync.PACKET_CODEC);

        PayloadTypeRegistry.playC2S().register(RequestSync.PACKET_ID, RequestSync.PACKET_CODEC);

        PayloadTypeRegistry.playC2S().register(Delete.PACKET_ID, Delete.PACKET_CODEC);
        PayloadTypeRegistry.playC2S().register(Report.PACKET_ID, Report.PACKET_CODEC);
        PayloadTypeRegistry.playC2S().register(Version.PACKET_ID, Version.PACKET_CODEC);

        PayloadTypeRegistry.playS2C().register(Delete.PACKET_ID, Delete.PACKET_CODEC);
        PayloadTypeRegistry.playS2C().register(Premium.PACKET_ID, Premium.PACKET_CODEC);


        ClientPlayNetworking.registerGlobalReceiver(Info.PACKET_ID, (payload, unused) -> Initializer.onDisplayInfoPacket(payload));
        ClientPlayNetworking.registerGlobalReceiver(Premium.PACKET_ID, (payload, unused) -> Initializer.onPremiumPacket(payload));
        ClientPlayNetworking.registerGlobalReceiver(Delete.PACKET_ID, (deletePacket, unused) -> Initializer.onDeletePacket(deletePacket));

        ClientPlayNetworking.registerGlobalReceiver(Sync.PACKET_ID, (payload, unused) -> Initializer.onSyncPacket(payload));

        WorldRenderEvents.AFTER_ENTITIES.register(context -> {
            Minecraft minecraft = Minecraft.getInstance();
            if (minecraft.level == null || minecraft.player == null) {
                return;
            }
            PoseStack matrices = context.matrices();
            Camera camera = context.gameRenderer().getMainCamera();
            World.render(matrices, camera);
        });


        ClientTickEvents.END_CLIENT_TICK.register(Initializer::onEndTick);

        ClientLifecycleEvents.CLIENT_STOPPING.register(minecraftClient -> Initializer.onStop());
    }

    @Override
    public void sendPacket(CustomPacketPayload packet) {
        ClientPlayNetworking.send(packet);
    }
}

File: fabric/src/main/kotlin/com/dreamdisplays/DreamDisplaysMod.kt
package com.dreamdisplays

import com.dreamdisplays.display.DisplayManager
import com.dreamdisplays.net.Packets
import com.dreamdisplays.render.ScreenRenderer
import net.fabricmc.api.ClientModInitializer
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientLifecycleEvents
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayConnectionEvents
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.fabricmc.fabric.api.client.rendering.v1.world.WorldRenderEvents
import net.fabricmc.fabric.api.networking.v1.PayloadTypeRegistry
import net.minecraft.client.Minecraft
import net.minecraft.network.protocol.common.custom.CustomPacketPayload

@Suppress("UNUSED")
class DreamDisplaysMod : ClientModInitializer, Mod {

    override fun onInitializeClient() {
        Initializer.onModInit(this)

        with(PayloadTypeRegistry.playS2C()) {
            register(Packets.Info.PACKET_ID, Packets.Info.PACKET_CODEC)
            register(Packets.Sync.PACKET_ID, Packets.Sync.PACKET_CODEC)
            register(Packets.Premium.PACKET_ID, Packets.Premium.PACKET_CODEC)
            register(Packets.Delete.PACKET_ID, Packets.Delete.PACKET_CODEC)
            register(Packets.DisplayEnabled.PACKET_ID, Packets.DisplayEnabled.PACKET_CODEC)
            register(Packets.ReportEnabled.PACKET_ID, Packets.ReportEnabled.PACKET_CODEC)
            register(Packets.ClearCache.PACKET_ID, Packets.ClearCache.PACKET_CODEC)
        }

        with(PayloadTypeRegistry.playC2S()) {
            register(Packets.Sync.PACKET_ID, Packets.Sync.PACKET_CODEC)
            register(Packets.RequestSync.PACKET_ID, Packets.RequestSync.PACKET_CODEC)
            register(Packets.Delete.PACKET_ID, Packets.Delete.PACKET_CODEC)
            register(Packets.Report.PACKET_ID, Packets.Report.PACKET_CODEC)
            register(Packets.Version.PACKET_ID, Packets.Version.PACKET_CODEC)
            register(Packets.SetVideo.PACKET_ID, Packets.SetVideo.PACKET_CODEC)
        }

        ClientPlayNetworking.registerGlobalReceiver(Packets.Info.PACKET_ID) { payload, _ ->
            Initializer.onDisplayInfoPacket(payload)
        }
        ClientPlayNetworking.registerGlobalReceiver(Packets.Premium.PACKET_ID) { payload, _ ->
            Initializer.onPremiumPacket(payload)
        }
        ClientPlayNetworking.registerGlobalReceiver(Packets.Delete.PACKET_ID) { payload, _ ->
            Initializer.onDeletePacket(payload)
        }
        ClientPlayNetworking.registerGlobalReceiver(Packets.DisplayEnabled.PACKET_ID) { payload, _ ->
            Initializer.onDisplayEnabledPacket(payload)
        }
        ClientPlayNetworking.registerGlobalReceiver(Packets.Sync.PACKET_ID) { payload, _ ->
            Initializer.onSyncPacket(payload)
        }
        ClientPlayNetworking.registerGlobalReceiver(Packets.ReportEnabled.PACKET_ID) { payload, _ ->
            Initializer.onReportEnabledPacket(payload)
        }
        ClientPlayNetworking.registerGlobalReceiver(Packets.ClearCache.PACKET_ID) { payload, _ ->
            Initializer.onClearCachePacket(payload)
        }

        WorldRenderEvents.AFTER_ENTITIES.register { context ->
            val mc = Minecraft.getInstance()
            if (mc.level != null && mc.player != null) {
                ScreenRenderer.render(context.matrices(), context.gameRenderer().mainCamera)
            }
        }

        ClientTickEvents.END_CLIENT_TICK.register { Initializer.onEndTick(it) }

        ClientPlayConnectionEvents.JOIN.register { _, _, client ->
            if (client.level != null && client.player != null) {
                val serverId = if (client.isLocalServer) "singleplayer"
                else client.currentServer?.ip ?: "unknown"
                DisplayManager.loadScreensForServer(serverId)
            }
        }

        ClientPlayConnectionEvents.DISCONNECT.register { _, _ ->
            DisplayManager.saveAllScreens()
            DisplayManager.unloadAll()
        }

        ClientLifecycleEvents.CLIENT_STOPPING.register { Initializer.onStop() }
    }

    override fun sendPacket(packet: CustomPacketPayload) {
        ClientPlayNetworking.send(packet)
    }
}
